### PR TITLE
Create intermediate models safely in case of cache miss

### PR DESCRIPTION
**Pull request by alllex**
When there is a cache miss for an intermediate model or a project metadata, we are running some work to create the value, store it in the cache, and then return it. This work is currently not protected against concurrent cache misses, which might happen, e.g. due to fetching models in parallel.

This PR protects the creation work with a `CalculatedValue`, ensuring it runs at most once.

### CODE REVIEW

**Comment by alllex** [comment]
@bot-gradle test without pts

**Comment by alllex** [comment]
This PR is currently stuck with a regression. I added a minimal `reproducer` test that demonstrates the problem:
https://github.com/gradle/gradle/blob/fdd1b525d2b67480936ca8b4dad6a9e50cf4c501/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiCompositeBuildsIntegrationTest.groovy#L56-L63

The symptom is that when project metadata (`LocalComponentGraphResolveState`) is deserialized from the project store (`ProjectMetadataController`), it somehow loses information. The reproducer demonstrates that plugins exported by an included build are no longer visible on the build script classpath.

I patched the code to demonstrate that it's exactly the project metadata serialization that is the problem:
https://github.com/gradle/gradle/blob/fdd1b525d2b67480936ca8b4dad6a9e50cf4c501/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt#L113-L122

**Comment by alllex** [comment]
After the discussion with @adammurdoch we decided to unblock this fix by special casing the `ProjectMetadataController` value loading.

More specifically, the problem was in the loss of dependency resolution information from the `LocalComponentGraphResolveState` with load-after-store semantics. While doing load-after-store for the intermediate models worked as expected.

When the component state is serialized, a good chunk of the state is discarded because it is not needed for the actual loads in subsequent cache hits. However, for the very first time we create the state, we need to preserve the dependency resolution info, so that tasks producing the project component artifacts are actually scheduled, and not ignored.

The temporary solution is to ensure the metadata controller avoids load-after-store the first time the project metadata is requested and returns the original value. This is safe in the sense that we do not run any user-provided code, and we also have additional memorization and synchronization around that call:
https://github.com/gradle/gradle/blob/42e894899a85a4beeb37b7bf28e7e9a6f11d859c/subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultBuildTreeLocalComponentProvider.java#L90

This is not ideal, but a sufficiently good solution for now.

A more rigorous approach would be to split the component state. One part would be loaded/created via the store, the other part that is required for correct task scheduling would not go through the store at all.

**Comment by alllex** [comment]
@bot-gradle test without pts

**Comment by alllex** [comment]
@bot-gradle test without pts

File: platforms/core-configuration/kotlin-dsl-tooling-builders/src/main/kotlin/org/gradle/kotlin/dsl/tooling/builders/internal/IsolatedProjectsSafeKotlinDslScriptsModelBuilder.kt [hunk 1]
```diff
@@ -294,9 +294,11 @@ fun isolatedScriptsModelFor(project: ProjectInternal): IsolatedScriptsModel {
     // TODO:isolated compute own classpaths
     val additionalClassPath = ClassPath.EMPTY
     val additionalSourcePath = ClassPath.EMPTY
-    val models = buildList {
-        addNotNull(buildScriptModelFor(project, additionalClassPath, additionalSourcePath))
-        this.addAll(precompiledScriptModelsFor(project))
+    val models = mutableListOf<IntermediateScriptModel>().apply {
```
Comment (alllex) [inline platforms/core-configuration/kotlin-dsl-tooling-builders/src/main/kotlin/org/gradle/kotlin/dsl/tooling/builders/internal/IsolatedProjectsSafeKotlinDslScriptsModelBuilder.kt:297]: Build manually until https://github.com/gradle/gradle/issues/28325 is fixed.

I tried using Guava, but it turns out this module does not have a direct dependency on it yet, so I'd avoid adding it just for this.

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiParallelModelQueryIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    @Rule
+    BlockingHttpServer server = new BlockingHttpServer()
+
+    def setup() {
+        server.start()
+    }
+
+    def "intermediate model is cached and reused for nested concurrent requests"() {
+        withSomeToolingModelBuilderPluginInBuildSrc("""
+            Thread.sleep(3000) // Simulate long-running builder to ensure overlap
```
Comment (adammurdoch) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiParallelModelQueryIntegrationTest.groovy:34]: Do we need this call to `sleep()`? It's usually pretty unreliable. The call to the blocking URL in the build action can move into the model builder instead to ensure the models are being built in parallel.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt [hunk 3]
```diff
@@ -93,33 +100,80 @@ abstract class ProjectStateStore<K, V>(
...
-            }
+        val address = loadOrCreateAddress(key, creator)
+        return readValue(key, address)
+    }
+
+    /**
+     * If value has to be created, the original value is returned without (de)serialization
+     *
+     * This is delicate API, because the original value is returned only for the first call of this function.
+     * As such, the call should probably be guarded with additional synchronization and memoization.
```
Comment (adammurdoch) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt:115]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt [hunk 4]
```diff
@@ -179,8 +181,13 @@ class DefaultConfigurationCache internal constructor(
         return intermediateModels.value.loadOrCreateIntermediateModel(identityPath, modelName, parameter, creator)
     }
 
+    // TODO:configuration - split the component state, such that information for dependency resolution does not have to go through the store
     override fun loadOrCreateProjectMetadata(identityPath: Path, creator: () -> LocalComponentGraphResolveState): LocalComponentGraphResolveState {
-        return projectMetadata.value.loadOrCreateValue(identityPath, creator)
+        // We are preserving the original value if it had to be created,
+        // because it carries information required by dependency resolution
+        // to ensure project artifacts are actually created the first time around.
+        // When the value is loaded from the store, the dependency information is lost.
```
Comment (adammurdoch) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt:189]: 👍

**Review by adammurdoch — CHANGES_REQUESTED** [review]

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/FetchCustomModelInParallel.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+        @Override
+        public SomeToolingModel execute(BuildController controller) {
+            awaitBlockingHttpServer(url); // Ensure each nested action is running in parallel
+            return controller.findModel(SomeToolingModel.class);
+        }
+    }
+
+    // CHECKSTYLE:OFF
+    private static void awaitBlockingHttpServer(String url0) {
```
[the remark this answers is not in the document]
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/FetchCustomModelInParallel.java:62]: That is what I initially wanted. However, the models are never built in parallel in this case due to a project lock:
https://github.com/gradle/gradle/blob/a3dc91fa697b5de5fd9c2f4ce13e1fbda561fa26/subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/DefaultToolingModelBuilderRegistry.java#L96

So I opted for blocking in the top-level build action, to make sure the models are _requested_ in parallel and to block as close to the model builders as possible. And the `sleep` is there to make sure the `creator()` call in the `ProjectStateStore` is taking long enough to observe multiple threads reaching that point.

In summary, the test is not about concurrent creation of models, but rather about duplicate re-creation in case of concurrent requests.

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiParallelModelQueryIntegrationTest.groovy (hunk 2 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiParallelModelQueryIntegrationTest.groovy:34]: The individual models are not built in parallel due to a project lock. That's why I am catching it with the `sleep`.
See the other comment for details.

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/FetchCustomModelInParallel.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/FetchCustomModelInParallel.java]: After taking another look, I think having the blocking server there does not actually make the test any better.

We have separate tests that ensure nested requests run in parallel, so we don't need to block there. What we do need is to make sure the first creator of the model runs long enough so that the second request catches up. Then we validate that despite this, we have created only a single model (the builder executed once).

I simplified the test.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt [hunk 6]
```diff
@@ -190,22 +197,38 @@ class DefaultConfigurationCache internal constructor(
...
+    fun collectProjectUsage(): ProjectUsage {
+        val reusedProjects = mutableSetOf<Path>()
+        val updatedProjects = mutableSetOf<Path>()
+        intermediateModels.value.visitProjects(reusedProjects::add, updatedProjects::add)
+        projectMetadata.value.visitProjects(reusedProjects::add) { }
+        return ProjectUsage(reusedProjects, updatedProjects)
+    }
+
+    private
+    data class ProjectUsage(val reused: Set<Path>, val updated: Set<Path>)
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt:220]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt [hunk 7]
```diff
@@ -190,22 +197,38 @@ class DefaultConfigurationCache internal constructor(
...
+
+    private
+    fun commitCacheEntry(reusedProjects: Set<Path>) {
+        store.useForStore { layout ->
+            writeConfigurationCacheFingerprint(layout, reusedProjects)
+            val usedModels = intermediateModels.value.collectAccessedValues()
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt:226]: 💅 It makes sense to extract these `intermediateModels.value` and `projectMetadata.value` expressions to a single place. It could be in a separate PR though.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt [hunk 8]
```diff
@@ -93,33 +100,80 @@ abstract class ProjectStateStore<K, V>(
...
+     *
+     * Prefer [loadOrCreateValue] unless there is a special reason.
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt:117]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt [hunk 9]
```diff
@@ -93,33 +100,80 @@ abstract class ProjectStateStore<K, V>(
...
+     */
+    fun loadOrCreateOriginalValue(key: K, creator: () -> V): V {
+        val originalValueCapture = AtomicReference<V>()
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt:120]: 💅 It seems a regular `var` would be simpler since there's no sharing involved?

```kotlin
var originalValueCapture: V? = null
```

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt [hunk 10]
```diff
@@ -93,33 +100,80 @@ abstract class ProjectStateStore<K, V>(
...
+    }
+
+    private
+    fun loadOrCreateAddress(key: K, creator: () -> V): BlockAddress {
+        val valueContainer = currentValues.computeIfAbsent(key) { k ->
+            createContainer(k, creator)
+        }
+
+        // Calculate the value after adding the entry to the map, so that the value container can take care of thread synchronization
+        valueContainer.finalizeIfNotAlready()
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt:142]: 👍

File: subprojects/core/src/main/java/org/gradle/internal/model/CalculatedValueContainerFactory.java [hunk 11]
```diff
@@ -49,6 +52,11 @@ public <T, S extends ValueCalculator<? extends T>> CalculatedValueContainer<T, S
         return new CalculatedValueContainer<>(displayName, new SupplierBackedCalculator<>(supplier), projectLeaseRegistry, globalContext);
     }
 
+    /**
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/internal/model/CalculatedValueContainerFactory.java:55]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt (hunk 7 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt:226]: Good idea! It's relatively rare to see `= lazy { ... }` compared to `by lazy { ... }`, but Kotlin does not seem to provide a clean mechanism to check initialization-ess of the `by lazy` property, like [it does for the `lateinit`s](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/is-initialized.html).

A follow-up PR would be better for this

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt (hunk 9 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt:120]: Taking the current calls to this method into account, I would agree.

At the same time, locally, the capturing lambda could still be called from a different thread if it happens to call `finalizeIfNotAlready` before the current thread. And even though there are probably enough happens-before relations to guarantee it would work, I still prefer to keep the concurrent nature of this capturing more explicit.

On a related note, there is something "shady" about Kotlin wrapping local variables in `Ref<>` and friends to make them mutable in lambdas, and I don't like it 😅 It's over that line for me, where Kotlin is too implicit

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt (hunk 7 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultConfigurationCache.kt:226]: Here is the PR:
https://github.com/gradle/gradle/pull/28431

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt (hunk 9 above)
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt:120]: > the capturing lambda could still be called from a different thread if it happens to call finalizeIfNotAlready before the current thread. 

Right 👍
